community/bolt/src/main/java/org/neo4j/bolt/v1/packstream/utf8/SunMiscUTF8Encoder.java

```diff
@@ -0,0 +1,84 @@
+package org.neo4j.bolt.v1.packstream.utf8;
+
+import sun.nio.cs.ArrayEncoder;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.StandardCharsets;
+
+public class SunMiscUTF8Encoder implements UTF8Encoder
+{
+    private final CharsetEncoder charsetEncoder = StandardCharsets.UTF_8.newEncoder();
+    private final ArrayEncoder arrayEnc = (ArrayEncoder) charsetEncoder;
+
+    private final byte[] out = new byte[1024 * 16];
+    private final ByteBuffer outBuf = ByteBuffer.wrap( out );
+
+    private final MethodHandle getCharArray;
+
+    private final UTF8Encoder fallbackEncoder = new VanillaUTF8Encoder();
+    private final int fallbackAtStringLength =
+            (int)(out.length / charsetEncoder.averageBytesPerChar());
+
+    /** Instantiated through reflection in {@link org.neo4j.bolt.v1.packstream.utf8.UTF8Encoder */
+    public SunMiscUTF8Encoder()
+    {
+        getCharArray = charArrayGetter();
+    }
+
+    @Override
+    public ByteBuffer encode( String input )
+    {
+        try
+        {
+            // If it's unlikely we will fit the encoded data, just use stdlib encoder
+            if( input.length() > fallbackAtStringLength )
+            {
+                return fallbackEncoder.encode( input );
+            }
+
+            char[] rawChars = (char[]) getCharArray.invoke( input );
+            int len = this.arrayEnc.encode( rawChars, 0, rawChars.length, out );
+
+            if( len == -1 )
+            {
+                return fallbackEncoder.encode( input );
+            }
+
+            outBuf.position(0);
+            outBuf.limit(len);
+            return outBuf;
+        }
+        catch( ArrayIndexOutOfBoundsException e )
+        {
+            // This happens when we can't fit the encoded string.
+            // We try and avoid this altogether by falling back to the
+            // vanilla encoder if the string looks like it'll not fit -
+            // but this is probabilistic since we don't know until we've encoded.
+            // So, if our guess is wrong, we fall back here instead.
+            return fallbackEncoder.encode( input );
+        }
+        catch ( Throwable e )
+        {
+            throw new AssertionError( e );
+        }
+    }
+
+    private MethodHandle charArrayGetter()
+    {
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+        try
+        {
+            Field value = String.class.getDeclaredField( "value" );
+            value.setAccessible( true );
+            return lookup.unreflectGetter( value );
+        }
```

Comment:
IMO we should look into dropping `java.lang.String` and using some custom thing instead. I may be underestimating our usage of java string functionality internally, but I don't think it's excessive, and the benefits to controlling string buffers directly would be quite nice. All big strings in the stores are utf-8, and the ones that use the custom encodings could be modified to decode directly to utf-8 onto the network. Likewise, basic predicates could be done without encoding/decoding - and for more advanced things like uppercase/lowercase/ordering, the java implementation is incorrect anyway. I guess regex is the big questionmark..

Anyway, this solves the immediate problem. Lots of opportunities to explore ahead!